osu.Game/Online/SignalRUnionWorkaroundResolver.cs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using MessagePack;
+using MessagePack.Formatters;
+using MessagePack.Resolvers;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Multiplayer.MatchTypes.TeamVersus;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// Handles SignalR being unable to comprehend [Union] types correctly by redirecting to a known base (union) type.
+    /// See https://github.com/dotnet/aspnetcore/issues/7298.
+    /// </summary>
+    public class SignalRUnionWorkaroundResolver : IFormatterResolver
+    {
+        public static readonly MessagePackSerializerOptions OPTIONS =
+            MessagePackSerializerOptions.Standard.WithResolver(new SignalRUnionWorkaroundResolver());
+
+        private static readonly Dictionary<Type, IMessagePackFormatter> formatter_map = new Dictionary<Type, IMessagePackFormatter>
+        {
+            { typeof(TeamVersusUserState), new TypeRedirectingFormatter<TeamVersusUserState, MatchUserState>() },
+            { typeof(TeamVersusRoomState), new TypeRedirectingFormatter<TeamVersusRoomState, MatchRoomState>() },
+            { typeof(ChangeTeamRequest), new TypeRedirectingFormatter<ChangeTeamRequest, MatchUserRequest>() },
+
+            // These should not be required. The fallback should work. But something is weird with the way caching is done.
```

Comment:
If someone is to ignore my comment and look into this further, it's something to do with the internal caching mechanism and the order/time at which `StandardResolver.Instance.GetFormatter<T>` is called. Or maybe which class it's called from.